src/types/timestamp.rs

```diff
@@ -354,6 +354,12 @@ impl TimestampBuilder {
     where
         D: Datelike + Timelike + Debug,
     {
+        if self.year == 0 || self.year > 9999 {
+            return illegal_operation(format!(
+                "Timestamp year '{}' out of range (1-9999)",
+                self.year
+            ));
+        }
```

Comment:
PR tour: this fixed several failures that were surfaced by `ion-tests`. I probably should've waited and done it in a follow-on PR, but I got excited.

Reply:
The annotation question above drove me to read through the ANTLR grammar for text encoding, and I noted the construct used there to enforce this limit: 
```plain

fragment
YEAR
    : '000'                     [1-9]
    | '00'            [1-9]     DEC_DIGIT
    | '0'   [1-9]     DEC_DIGIT DEC_DIGIT
    | [1-9] DEC_DIGIT DEC_DIGIT DEC_DIGIT
    ;
```